[Kaian] iP

caipng opened the pull request:
# Duke
> Don't believe everything you read on the internet just because there's a picture with a quote next to it. - Albert Einstein ([source](https://www.pinterest.com/pin/377035800043983277/))

Duke frees your mind of having to remember things you need to do (not by deleting your tasks). It's,
- text-based
- easy to learn (even though its text based)
- ~FAST~ *SUPER* FAST to use (tested on a 24 core processor)

All you need to do is,
1. download it from [here](https://github.com/caipng/ip)
2. double-click it (figure out where yourself)
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**! ^

^ for the first 24 minutes, chargeable at $10 per task thereafter

Features:
- [X] Managing tasks 
- [ ] Managing deadlines (that's on your self discipline) 
- [ ] Reminders (purchasable in the program for an additional nominal fee; cmd: `buy reminders <card details>`)
---
If you are a Java programmer (sorry), you can use it to practice Java too. Here's the `main` method (don't plagiarise please):
```java
public static void main(String[] args) {
    try {
        Duke duke = new Duke();
        duke.run();
    } catch (Exception e) {
        e.printStackTrace();
    }
}
```

Commit:
Add Gradle support

caipng committed:
Add Java class files to .gitignore

caipng committed:
Setup Duke package

caipng committed:
Fix ordering of modifiers in Duke class

caipng committed:
Add Duke.Commands package

caipng committed:
Setup basic functionality

caipng committed:
Add Duke.Todo package

caipng committed:
Add command for adding todos

caipng committed:
Add command for listing todos

caipng committed:
Add ability to mark a todo as done

caipng committed:
Add command to mark a todo as done

caipng committed:
Rename todo to task

caipng committed:
Add todo class

caipng committed:
Add deadline and event classes

caipng committed:
Enable commands to match against multiple keywords

caipng committed:
Extend add todo command to work with deadline and event

caipng committed:
Update text UI test

caipng committed:
Use custom exceptions to handle unknown commands

caipng committed:
Add empty description exception

caipng committed:
Extend EmptyDescriptionException to InvalidTaskException

caipng committed:
Throw exceptions for empty task dates

caipng committed:
Add error handling for done task command

caipng committed:
Add remove task feature

caipng committed:
Use enums for task types

caipng committed:
Ignore case in task date regex

caipng committed:
Add task storage feature

caipng committed:
Store dates as LocalDate instead of String

caipng committed:
Merge branch 'branch-Level-8'

caipng committed:
Merge branch 'branch-Level-7'

caipng committed:
Extract Ui and Parser class

caipng committed:
Refactor UserInput and Parser class

caipng committed:
Add test for parser

caipng committed:
Add storage interface

caipng committed:
Add tests for task list

caipng committed:
Add Manifest file

caipng committed:
Add find task by keyword feature

caipng committed:
Add JavaDocs

caipng committed:
Format code to follow coding standard

caipng committed:
Merge branch 'branch-Level-9'

caipng committed:
Merge branch 'branch-A-Javadoc'

caipng committed:
Merge branch 'branch-A-CodingStandard'

felix-ong commented on src/main/java/Duke/Task/InvalidTaskException.java:
```diff
@@ -0,0 +1,15 @@
+package Duke.Task;
+
+import Duke.DukeException;
+
```
Maybe it would be good to include a header comment here to describe when this exception is encountered since it might not be clear what an invalid task is?

felix-ong commented on src/main/java/Duke/Task/TaskListTest.java:
```diff
@@ -0,0 +1,51 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.fail;
```
```suggestion
import Duke.Storage.Storage;

import java.io.IOException;
import java.util.List;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.fail;
```
Maybe you can put the org.junit.jupiter.api.Test import statement on a separate line for clarity?

felix-ong commented on README.md:
```diff
@@ -1,19 +1,24 @@
-3. After that, locate the `src/main/java/Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
+3. After that, locate the `src/main/java/Duke/Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code
+   editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the
+   below as the output:
```
I like how you adhered to the line length limit for the README as well

felix-ong commented on src/main/java/Duke/Storage/TaskStorage.java:
```diff
@@ -0,0 +1,166 @@
+            throw new FileFormatException("Expected 4 values, got " + splitData.length);
+        }
+
+        int completionStatus;
```
```suggestion
        int completionStatus = 0;
```
The coding standards mentioned that variables should be initialized where they are declared so maybe you could do this?

felix-ong commented on src/main/java/Duke/Task/TaskList.java:
```diff
@@ -0,0 +1,79 @@
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
```
Maybe it would be good to include header comments for this class and its methods since it is an integral part of this program?

felix-ong reviewed (commented):
LGTM. Overall, the code is very well written and easy to understand. I struggled to find any coding standards violations. :sweat_smile: Besides including more header comments, I do not think that there is anything else that can be improved with regards to coding standards. Good job! :satisfied:

ganhongyao commented on src/main/java/Duke/Commands/Command.java:
```diff
@@ -0,0 +1,80 @@
+     * @param input The user input that triggered this command.
+     * @throws DukeException If a DukeException was thrown when running the command.
+     */
+    abstract public void run(Duke duke, UserInput input) throws DukeException;
```
Should the order of method modifiers be changed such that it follows the JLS format? i.e. 

```
<access> static abstract synchronized <unusual> final native
```

I noticed this in a few other places too.

```suggestion
    public abstract void run(Duke duke, UserInput input) throws DukeException;
```

ganhongyao commented on src/main/java/Duke/Duke.java:
```diff
@@ -0,0 +1,84 @@
+    public Duke(String todoStoragePath) throws IOException {
+        Storage<Task> taskStorage = new TaskStorage(todoStoragePath);
+        try {
+            this.taskList = new TaskList(taskStorage);
```
Should we avoid the unnecessary use of the `this` keyword? See https://se-education.org/guides/conventions/java/index.html#variables.

I noticed this in a few other places too.

```suggestion
            taskList = new TaskList(taskStorage);
```

ganhongyao commented on src/main/java/Duke/Commands/AddTaskCommand.java:
```diff
@@ -0,0 +1,89 @@
+package Duke.Commands;
```
Should package names be in all lower-case instead?

ganhongyao commented on src/main/java/Duke/Storage/TaskStorage.java:
```diff
@@ -0,0 +1,166 @@
+        if (task instanceof Deadline) {
+            date = ((Deadline) task).getEndDate().toString();
+        }
+        if (task instanceof Event) {
+            date = ((Event) task).getAt().toString();
+        }
```
Would it make sense to use `else if` here since a `Task` cannot be a `Deadline` and `Event` at the same time?

ganhongyao reviewed (commented):
I like how clean your code is that I had to pull out some advanced rules found [here](https://se-education.org/guides/conventions/java/index.html) else my review would look empty.

Nice 😜

caipng committed:
Merge branch 'add-gradle-support'

caipng committed:
Move tests to dedicated `test` source set

caipng committed:
Refactor `main` class

caipng committed:
Configure checkstyle

caipng committed:
Add JavaFX

caipng committed:
Implement GUI

caipng committed:
Add assertions

caipng committed:
Improve code quality

Some `switch` statements lack a `default` case. Also, the `parseTask` method is too long.

This is in violation of the code quality standards.

Let's
* Add a `default` case for all `switch` statements
* Extract method `parseCompletionStatus` from `parseTask`

This will improve the code quality.

caipng committed:
Merge pull request #2 from caipng/branch-A-Assertions

Add assertions

caipng committed:
Merge branch 'master' into branch-A-CodeQuality

caipng committed:
Merge pull request #1 from caipng/branch-A-CodeQuality

Improve code quality

caipng committed:
Add more aliases

Implementation of `AddTaskCommand` does not allow for aliases.

Users are only limited to `todo`, `deadline` and `event` keywords for adding tasks.

Let's change the implementation of `AddTaskCommand` to allow for aliases.

Adding such a functionality will provide more flexibility, especially for expert users.

caipng committed:
Improve GUI

caipng committed:
Change message personality

caipng committed:
Rename package names to use all lowercase

caipng committed:
Suppress unused return value warning

caipng committed:
Fix checkstyle errors

caipng committed:
Fix typo in DialogBox.fxml

caipng committed:
Add user guide

caipng committed:
Set theme jekyll-theme-slate

caipng committed:
Finalize for submission

caipng committed:
Add credits to README.md
